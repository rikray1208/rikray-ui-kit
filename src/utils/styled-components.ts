import * as styledComponents from 'styled-components';
import {ITheme} from "../theme/types"


const {
    default: styled,
    css,
    createGlobalStyle,
    keyframes,
    ThemeProvider,
} = styledComponents as styledComponents.ThemedStyledComponentsModule<ITheme>;

export { css, createGlobalStyle, keyframes, ThemeProvider };
export default styled;
