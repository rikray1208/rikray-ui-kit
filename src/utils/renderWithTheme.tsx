import {render} from "@testing-library/react";
import React from "react";
import {ITheme} from "../theme/types";
import {ThemeProvider} from "./styled-components";

export const renderWithTheme = (component: React.ReactNode, theme: ITheme) => {
    return render(
        <ThemeProvider theme={theme}>
            {component}
        </ThemeProvider>
    )
}
