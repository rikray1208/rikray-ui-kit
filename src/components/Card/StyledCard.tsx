import styled, {css} from "styled-components";
import {CardProps} from "./index";
import {setShape} from "../../utils/helpers";

type StyledProps = Pick<CardProps, 'shape'>

export const CardWrapper = styled.div<StyledProps>`
  ${({theme, shape}) => css`
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    
    width: fit-content;
    
    background: ${theme.Card.colorBgContainer};
    border: ${theme.Card.borderWidth} ${theme.Card.borderStyle} ${theme.Card.colorBorderBase};

    font-size: ${theme.Button["fontSize-base"] + 'rem'};
    line-height: ${theme.Button["lineHeight-base"] + 'rem'};
    font-family: ${theme.Button["fontFamily"]};
    color: ${theme.Card.colorTextBase};
    
    ${
            shape && setShape(shape, 'Card')}
    }
  `}
`
const commonStyles = styled.div`
  background: transparent;
  font-size: ${({theme}) => theme.Card["fontSize-lg"] + 'rem'};
  line-height: ${({theme}) => theme.Card["lineHeight-lg"] + 'rem'};
`
export const CardHeader = styled(commonStyles)`
  ${({theme}) => css`
    padding: ${theme.Card["padding-3"] + 'rem'};
    font-weight: ${theme.Card.fontWeightBold};
    border-bottom: ${theme.Card.borderWidth} ${theme.Card.borderStyle} ${theme.Card.colorBorderBase};
  `}
`
export const CardFooter = styled(commonStyles)`
  ${({theme}) => css`
    margin-top: auto;
    padding: ${theme.Card["padding-2"] + 'rem'};
    border-top: ${theme.Card.borderWidth} ${theme.Card.borderStyle} ${theme.Card.colorBorderBase};
  `}`
export const CardContent = styled.div`
    background: transparent;
    padding: ${({theme}) => theme.Card["padding-3"] + 'rem'};
`;

