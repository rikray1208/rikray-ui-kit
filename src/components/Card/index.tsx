import React, {FC, PropsWithChildren} from 'react';
import {Shape, Size} from "../types";
import {CardContent, CardWrapper, CardHeader, CardFooter} from "./StyledCard";
import {ThemeProvider} from "../../index";

export interface CardProps {
    header?: string;
    footer?: string;
    shape?: Shape
}
const Card: FC<PropsWithChildren<CardProps>> = ({children, ...props}) => {
    return (
        <ThemeProvider>
            <CardWrapper shape={props.shape}>
                {props.header && <CardHeader>{props.header}</CardHeader>}
                <CardContent>
                    {children}
                </CardContent>
                {props.footer && <CardFooter>{props.footer}</CardFooter>}
            </CardWrapper>
        </ThemeProvider>
    );
};

Card.defaultProps = {
    header: 'Card title',
    children: <p style={{margin: 0}}>Card content...</p>,
    shape: 'SemiRound'
}
export default Card;
