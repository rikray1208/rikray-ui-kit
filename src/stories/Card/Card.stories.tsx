import {ComponentMeta, ComponentStory} from "@storybook/react";
import React from "react";
import {ThemeProvider} from "../../index";
import Card from "../../components/Card";

const decorators = [
    (Story: any) => (
        <ThemeProvider>
            <Story />
        </ThemeProvider>
    ),
];

export default {
    title: 'Design System/Global/Card',
    component: Card,
    decorators: [
        (Story: any) => (
            <ThemeProvider>
               <div>
                   <Story />
               </div>
            </ThemeProvider>
        ),
    ]
} as ComponentMeta<typeof Card>;


const Template: ComponentStory<typeof Card> = (args) => <Card {...args} />;

export const DefaultCard = Template.bind({});
DefaultCard.args = {
    header: 'Card title',
    children: [<p style={{margin: ' 0 100px 0 0'}}>Card content...</p>,<p style={{margin: 0}}>Card content...</p>,<p style={{margin: 0}}>Card content...</p>,<p style={{margin: 0}}>Card content...</p>],
    shape: 'SemiRound'
}

