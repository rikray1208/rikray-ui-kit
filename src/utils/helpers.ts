import {Shape, Size} from "../components/types";
import {css} from "styled-components";
import React from "react";
import renderer from "react-test-renderer";
import {ComponentMap} from "../theme/types";

type Component = keyof  ComponentMap;

export function setShape(shape: Shape, Component: Component) {
    if (shape === 'Round')
        return css`border-radius: ${({theme}) => theme[Component]["borderRadius-lg"] + 'px'};`

    if (shape === 'SemiRound')
        return css`border-radius: ${({theme}) => theme[Component]["borderRadius-sm"] + 'px'};`

    if (shape === 'Rectangle')
        return css`border-radius: ${({theme}) => theme[Component]["borderRadius-xs"] + 'px'};`
}
export function setFontSize(size: Size, Component: Component) {
    if (size == 'Small')
        return css`
          font-size: ${({theme}) => theme[Component]["fontSize-sm"] + 'rem'};
          line-height: ${({theme}) => theme[Component]["lineHeight-sm"] + 'rem'};
        `

    if (size == 'Medium')
        return css`
          font-size: ${({theme}) => theme[Component]["fontSize-base"] + 'rem'};
          line-height: ${({theme}) => theme[Component]["lineHeight-base"] + 'rem'};
        `

    if (size == 'Large')
        return css`
          font-size: ${({theme}) => theme[Component]["fontSize-lg"] + 'rem'};
          line-height: ${({theme}) => theme[Component]["lineHeight-lg"] + 'rem'};
        `
}

export const assertReactElement = (element: React.ReactElement) => {
    const component = renderer.create(element);
    return component.toJSON();
};
