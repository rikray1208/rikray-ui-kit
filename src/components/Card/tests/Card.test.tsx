import {ThemeProvider} from "../../../index";
import {assertReactElement} from "../../../utils/helpers";
import React from "react";
import Card from "../index";

describe('Card', () => {
    it('should return a Card', () =>{
        const element = (
            <ThemeProvider><Card/></ThemeProvider>
        );
        expect(assertReactElement(element)).toMatchSnapshot();
    })

    it('should return a SemiRound Card', () =>{
        const element = (
            <ThemeProvider><Card shape={'SemiRound'}/></ThemeProvider>
        );
        expect(assertReactElement(element)).toMatchSnapshot();
    })

    it('should return a Round Card', () =>{
        const element = (
            <ThemeProvider><Card shape={'Round'}/></ThemeProvider>
        );
        expect(assertReactElement(element)).toMatchSnapshot();
    })

    it('should return a Rectangle Card', () =>{
        const element = (
            <ThemeProvider><Card shape={'Rectangle'}/></ThemeProvider>
        );
        expect(assertReactElement(element)).toMatchSnapshot();
    })
})
