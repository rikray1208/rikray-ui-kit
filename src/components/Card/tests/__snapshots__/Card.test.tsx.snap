// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should return a Card 1`] = `
<div
  className="sc-iBYQkv jmiuSv"
  shape="SemiRound"
>
  <div
    className="sc-ftTHYK sc-pyfCe hMniyA dCYhwO"
  >
    Card title
  </div>
  <div
    className="sc-kDvujY jbOUMt"
  >
    <p
      style={
        {
          "margin": 0,
        }
      }
    >
      Card content...
    </p>
  </div>
</div>
`;

exports[`Card should return a Rectangle Card 1`] = `
<div
  className="sc-iBYQkv bfapFn"
  shape="Rectangle"
>
  <div
    className="sc-ftTHYK sc-pyfCe hMniyA dCYhwO"
  >
    Card title
  </div>
  <div
    className="sc-kDvujY jbOUMt"
  >
    <p
      style={
        {
          "margin": 0,
        }
      }
    >
      Card content...
    </p>
  </div>
</div>
`;

exports[`Card should return a Round Card 1`] = `
<div
  className="sc-iBYQkv gZDPuE"
  shape="Round"
>
  <div
    className="sc-ftTHYK sc-pyfCe hMniyA dCYhwO"
  >
    Card title
  </div>
  <div
    className="sc-kDvujY jbOUMt"
  >
    <p
      style={
        {
          "margin": 0,
        }
      }
    >
      Card content...
    </p>
  </div>
</div>
`;

exports[`Card should return a SemiRound Card 1`] = `
<div
  className="sc-iBYQkv jmiuSv"
  shape="SemiRound"
>
  <div
    className="sc-ftTHYK sc-pyfCe hMniyA dCYhwO"
  >
    Card title
  </div>
  <div
    className="sc-kDvujY jbOUMt"
  >
    <p
      style={
        {
          "margin": 0,
        }
      }
    >
      Card content...
    </p>
  </div>
</div>
`;
